Improve amcrest error handling and bump amcrest package to 1.5.3

pnbruckner opened the pull request:
## Description:
This component has been plagued by misleading network errors for some time. For the most part they have been due to improperly retrying network communications at the urllib3 level when Digest Authentication is being used, but urllib3 does not know how to properly handle that protocol, which is implemented at the requests level. This update is aimed at solving those problems.

Also, if a configured camera cannot be communicated with (e.g., it is temporarily turned off or there are intermittent network problems), the log is flooded with warnings and errors. The other major feature of this update is monitoring those errors and when they get too bad, taking the camera "offline" (i.e., making it unavailable), and going into a recovery mode where the camera is periodically "pinged" to see if/when it comes back "online." During this recovery mode normal updates and operations are suspended.

Lastly, the snapshot command used to send "?channel=0" or "?channel=1" depending on the resolution configuration variable. This was due to a misunderstanding of how the snapshot command works. Also, depending on camera model and firmware version, some can accept channel 0 (which used to be the default), whereas some now require channel 1. But all seem to default to the correct (and only) video channel if the "?channel=x" URL parameter is simply not sent. This is what the snapshot command will now do to make it compatible with more models and firmware versions.

### List of changes
amcrest package update fixes command retry, especially with Digest Authentication, and allows sending snapshot command without channel parameter.

Get rid of persistent_notification.

Errors at startup, other than login errors, are no longer fatal.

Display debug messages about how many times an error has occurred in a row.

Remove initial communications test. If camera is off line at startup this just delays setup, and hence HA startup time.

Handle urllib3 errors when getting data from commands that were sent with stream=True.

If errors occur during camera update, try repeating until it works or the camera is determined to be off line.

Drop channel parameter in snapshot command which allows camera to use its default channel, which is different in different camera models and firmware versions.

Make entities unavailable if too many errors occur in a row.

Add new configuration variables to control how many errors in a row should be interpreted as camera being offline, and how frequently to "ping" camera to see when it becomes available again.

Add online binary_sensor option to indicate if camera is available (i.e., responding to commands.)

**Related issue (if applicable):** fixes #20700

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#9554

## Example entry for `configuration.yaml` (if applicable):
```yaml
amcrest:
  - host: IP_ADDRESS_CAMERA
    username: YOUR_USERNAME
    password: YOUR_PASSWORD
    max_errors: 5
    recheck_interval:
        minutes: 2
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

pnbruckner committed:
Improve amcrest error handling and bump amcrest package to 1.5.3

amcrest package update fixes command retry, especially with Digest Authentication, and allows sending snapshot command without channel parameter.

Get rid of persistent_notification.

Errors at startup, other than login errors, are no longer fatal.

Display debug messages about how many times an error has occurred in a row.

Remove initial communications test. If camera is off line at startup this just delays the component setup.

Handle urllib3 errors when getting data from commands that were sent with stream=True.

If errors occur during camera update, try repeating until it works or the camera is determined to be off line.

Drop channel parameter in snapshot command which allows camera to use its default channel, which is different in different camera models and firmware versions.

Make entities unavailable if too many errors occur in a row.

Add new configuration variables to control how many errors in a row should be interpreted as camera being offline, and how frequently to "ping" camera to see when it becomes available again.

Add online binary_sensor option to indicate if camera is available (i.e., responding to commands.)

balloob commented:
For future PRs, please focus on smaller PRs. It's easier to review.

balloob commented on homeassistant/components/amcrest/__init__.py:
```diff
@@ -105,6 +115,10 @@ def _has_unique_names(devices):
                     _deprecated_sensor_values),
         vol.Optional(CONF_SWITCHES):
             vol.All(cv.ensure_list, [vol.In(SWITCHES)]),
+        vol.Optional(CONF_MAX_ERRORS, default=DEFAULT_MAX_ERRORS):
```
I don't think that these should be configurable. We should pick some decent defaults that work for the device and stick with it.

pnbruckner replied:
Done.

Unfortunately, depending on configuration, the rate at which errors can happen can vary widely (i.e., with more sensors configured, services called, etc, there will be more frequent communication.) It might make sense to measure the time period over which errors continue to happen and if it becomes too long, _then_ make the device unavailable. We'll go with this for now and I'll monitor how it goes. Might need to switch to time based error monitoring instead of count based in a follow-up PR.

balloob commented on homeassistant/components/amcrest/__init__.py:
```diff
@@ -117,31 +131,90 @@ def _has_unique_names(devices):
+    hass.data.setdefault(DATA_AMCREST, {DEVICES: {}, CAMERAS: []})
+
+    # pylint: disable=too-many-ancestors
+    class AmcrestChecker(Http):
```
Please define this outside of the setup method.

pnbruckner replied:
Done.

balloob commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -160,6 +173,14 @@ def __init__(self, name, device, ffmpeg):
+
+        False if entity pushes its state to HA.
+        """
+        return self._api.available and not self._update_succeeded
```
Home Assistant does not support should_poll to be dynamically changed. It's either polled or it's not.

pnbruckner replied:
Ok, changed the implementation.

balloob reviewed (approved):
Looks great. Few small comments. Ok to merge when addressed !

pnbruckner commented:
> For future PRs, please focus on smaller PRs. It's easier to review.
I will certainly try to keep that in mind.

pnbruckner committed:
Update per review comments

Remove max_errors and recheck_interval configuration variables and used fixed values instead.

Move definition of AmcrestChecker class to module level.

Change should_poll in camera.py to return a fixed value of True and move logic to update method.

CLARENNE-Q commented:
Invalid config for [amcrest]: [max_errors] is an invalid option for [amcrest]. Check: amcrest->amcrest->0->max_errors. (See /config/configuration.yaml, line 50). Please check the docs at https://home-assistant.io/integrations/amcrest/

On Home Assistant 0.103.0
